### PR TITLE
add API.md

### PR DESCRIPTION
**Pull request by nomad605dis**
Create API.md for GPIO Expander.

### CODE REVIEW

File: API.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+## `class GpioExpander`
+
+Create an object `GpioExpander` to communicate with a particular GpioExpander boards.
+
+### `GpioExpander(i2c_address=42)`
+
+The constructor for GpioExpander class. The `i2c_address` specifies the board
+I²C address which is `42` factory-default but can be changed programmatically.
+
+### `pin_mode(pin: int, value: int) -> None`
```
Comment (AlexErAmp) [inline API.md]: Change` "value" to "mode" in the parameter list.
```suggestion
### `pin_mode(pin: int, mode: int) -> None`
```

File: API.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+Configures state the `pin`:
```
Comment (AlexErAmp) [inline API.md]:
```suggestion
Configures the operation `mode` of the `pin`:
```

**Review by AlexErAmp — CHANGES_REQUESTED** [review]

File: API.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+### `analog_read(pin: int) -> float`
+
+Reads the analog value from one pin of board. Return value in range from `0` to `1.0`.
+
+### `digital_read(pin: int) -> int`
+
+Reads the bit value from one pin of board. Return `0` or `1` value.
+
+### `digital_write(pin: int) -> None`
```
Comment (yurybotov) [inline API.md]: ### `digital_write(pin: int, value: int) -> None`

**Review by yurybotov — CHANGES_REQUESTED** [review]
Look comment in text (digital_write)

File: API.md (hunk 3 above)
Comment (nomad605dis) [reply API.md]: I didn’t notice, thanks.

**Comment by nomad605dis** [comment]
Applied proposed changes.

**Review by AlexErAmp — APPROVED** [review]
LGTM

File: API.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
+# GPIO Expander Pi API
+
+## `class GpioExpander`
+
+Create an object `GpioExpander` to communicate with a particular GpioExpander boards.
```
Comment (nkrkv) [inline API.md]:
```suggestion
Create an object `GpioExpander` to communicate with a particular board based on Amperka GPIO expander. The end-user libraries often wrap (or inherit) this class to provide domain-specific high level hardware interaction methods.
```

File: API.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+### `GpioExpander(i2c_address=42)`
+
+The constructor for GpioExpander class. The `i2c_address` specifies the board
```
Comment (nkrkv) [inline API.md]:
```suggestion
Constructor. The `i2c_address` specifies the board
```

File: API.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+I²C address which is `42` factory-default but can be changed programmatically.
+
+### `pin_mode(pin: int, mode: int) -> None`
+
+Configures the operation `mode` of the `pin`:
+
+- INPUT
+- INPUT_PULLUP
+- INPUT_PULLDOWN
+- OUTPUT
```
Comment (nkrkv) [inline API.md]: What’s this? True enum, semi-enum, globals? Where are they defined? Use fully qualified names, please.

File: API.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+## `class GpioExpander`
+
+Create an object `GpioExpander` to communicate with a particular GpioExpander boards.
+
+### `GpioExpander(i2c_address=42)`
+
+The constructor for GpioExpander class. The `i2c_address` specifies the board
+I²C address which is `42` factory-default but can be changed programmatically.
+
+### `pin_mode(pin: int, mode: int) -> None`
```
Comment (nkrkv) [inline API.md]: `pin: int` ← what if I want to address `A3`? What `int` should I pass here?

File: API.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+Reads the analog value from one pin of board. Return value in range from `0` to `1.0`.
+
+### `digital_read(pin: int) -> int`
+
+Reads the bit value from one pin of board. Return `0` or `1` value.
+
+### `digital_write(pin: int, value: int) -> None`
+
+Write `1` or `0` value to a `pin`.
```
Comment (nkrkv) [inline API.md]: Should use booleans, not integers

File: API.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+### `analog_write(pin: int, value: int) -> None`
+
+Writes an analog value from `0` to `255` (PWM wave) to a `pin`.
```
Comment (nkrkv) [inline API.md]: Inconsistency: should be “from 0.0 to 1.0”, not the “0..255” cargo cult

**Review by nkrkv — CHANGES_REQUESTED** [review]
Run the file through prettier with `--prose-wrap=never`

**Review by yurybotov — APPROVED** [review]
ok

File: API.md (hunk 7 above)
Comment (nomad605dis) [reply API.md]: All boards based on the GPIO Expander have pins without alphabetic numbers.

File: API.md (hunk 6 above)
Comment (nomad605dis) [reply API.md]: Can we create a separate class `GPIO` where these constants will be static members and put it in a separate module?

File: API.md (hunk 8 above)
Comment (nomad605dis) [reply API.md]: We suggest adding to the GPIO class, which will contain all the constants, two static members: `HIGH` and `LOW` equal `True` and `False`, respectively. Then `value` in `digital_write` and the return value `digital_read` is changed to bool.

File: API.md (hunk 9 above)
Comment (nomad605dis) [reply API.md]: Done

**Review by nomad605dis — COMMENTED** [review]
Edit API.md and offer new additions.

File: API.md (hunk 6 above)
Comment (AlexErAmp) [reply API.md]: I think we can add these constants as static members directly to the GpioExpander class. Then we can call the `pin_mode` method as:
`obj.pin_mode(pin, GpioExpander.INPUT)`

File: API.md (hunk 7 above)
Comment (nkrkv) [reply API.md]: Ah, OK

File: API.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+Reads the bit value from one pin of board. Return `GPIO.HIGH` or `GPIO.LOW` value. This values defined as `True` and `False` respectively.
+
+### `digital_write(pin: int, value: bool) -> None`
+
+Write `GPIO.HIGH` or `GPIO.LOW` value to a `pin`. This values defined as `True` and `False` respectively.
+
+### `analog_write(pin: int, value: float) -> None`
+
+Writes an analog value from `0` to `1.0` (PWM wave) to a `pin`.
```
Comment (nkrkv) [inline API.md]:
```suggestion
Outputs PMW signal on a `pin`. The `value` argument specifies the duty cycle and should be in the range from `0.0` to `1.0`
```

File: API.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+### `change_address(new_address: int) -> None`
+
+Changes the I²C address of the module. The change is in effect only while the board is powered on. If you want to save it permanently call the `save_address` method.
+
+### `save_address() -> None`
+
+Permanently saves the current board I²C address.
+
+### `pwm_freq(value: int) -> None`
```
Comment (nkrkv) [inline API.md]:
```suggestion
### `set_pwm_freq(value: int) -> None`
```

File: API.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+Sets the PWM frequency for the analog output.
```
Comment (nkrkv) [inline API.md]: Information about valid (and sane) values of `value` is missing. Can I specify 12345678 here to output the signal with exactly 12345678 herz frequency? I doubt so.

**Review by nkrkv — CHANGES_REQUESTED** [review]

File: API.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+Constructor. The `i2c_address` specifies the board I²C address which is `42` factory-default but can be changed programmatically.
+
+### `pin_mode(pin: int, mode: int) -> None`
+
+Configures the operation `mode` of the `pin`:
+
+- GPIO.INPUT
+- GPIO.INPUT_PULLUP
+- GPIO.INPUT_PULLDOWN
+- GPIO.OUTPUT
```
Comment (nkrkv) [inline API.md]: Why `GPIO`? Looks too general. One of possible pythonic ways:

```py
from enum import Enum

class PinMode(Enum):
    INPUT = 1   #  use enum.auto() if the end-value “1” does not matter much
    INPUT_PULLUP = 2
    INPUT_PULLDOWN = 123
    OUTPUT = 3
```

**Review by yurybotov — APPROVED** [review]
OK

File: API.md (hunk 8 above)
Comment (nkrkv) [reply API.md]: Why so complicated? Just `True` and `False` is fine

File: API.md (hunk 13 above)
Comment (nomad605dis) [reply API.md]: Done

File: API.md (hunk 12 above)
Comment (nomad605dis) [reply API.md]: Done

**Review by nomad605dis — COMMENTED** [review]
Edit API.md

File: API.md [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+- PinMode.INPUT_PULLDOWN
+- PinMode.OUTPUT
+
+### `analog_read(pin: int) -> float`
+
+Reads the analog value from one pin of board. Return value in range from `0` to `1.0`.
+
+### `digital_read(pin: int) -> bool`
+
+Reads the bit value from one pin of board. Return `GPIO.HIGH` or `GPIO.LOW` value. This values defined as `True` and `False` respectively.
```
Comment (nkrkv) [inline API.md]:
```suggestion
Reads the bit value from one pin of the board. Return `GPIO.HIGH` or `GPIO.LOW` value. These values defined as `True` and `False` respectively.
```

So… why `GPIO.HIGH` and `GPIO.LOW` are ever required? Why can’t they be just `True` or `False`. OK, Arduino API does use HIGH and LOW, but this is _not_ Arduino and the de-facto standard `gpiozero` library does not use HIGH/LOW, etc.

Suggest KISS: just use `bool`.

File: API.md [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+### `digital_write(pin: int, value: bool) -> None`
+
+Write `GPIO.HIGH` or `GPIO.LOW` value to a `pin`. This values defined as `True` and `False` respectively.
```
Comment (nkrkv) [inline API.md]: Ditto

**Review by nkrkv — CHANGES_REQUESTED** [review]

File: API.md [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+Changes the I²C address of the module. The change is in effect only while the board is powered on. If you want to save it permanently call the `save_address` method.
+
+### `save_address() -> None`
+
+Permanently saves the current board I²C address.
+
+### `set_pwm_freq(freq: int) -> None`
+
+Sets the PWM frequency for the analog output. The `freq` must be in range from `30`(Hz) to `65535`(Hz).
```
Comment (nkrkv) [inline API.md]:
```suggestion
Sets a PWM frequency (in hertz) for all PWM outputs of the GPIO expander. The `freq` argument should be in the range from `30` to `65535`.
```

**Review by nkrkv — CHANGES_REQUESTED** [review]

File: API.md (hunk 14 above)
Comment (nomad605dis) [reply API.md]: Done

File: API.md (hunk 15 above)
Comment (nomad605dis) [reply API.md]: Done

File: API.md (hunk 16 above)
Comment (nomad605dis) [reply API.md]: Done

**Review by nomad605dis — COMMENTED** [review]
Applied proposed changes.

File: API.md [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+- PinMode.INPUT_PULLDOWN
+- PinMode.OUTPUT
+
+### `analog_read(pin: int) -> float`
+
+Reads the analog value from one pin of board. Return value in range from `0.0` to `1.0`.
+
+### `digital_read(pin: int) -> bool`
+
+Reads the bit value from one pin of board. Return `True` or `False` value.
```
Comment (nkrkv) [inline API.md]:
```suggestion
Reads a digital signal value from a pin of the board. Returns `True` or `False`.
```

File: API.md [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+### `digital_write(pin: int, value: bool) -> None`
+
+Write `True` or `False` value to a `pin`.
```
Comment (nkrkv) [inline API.md]:
```suggestion
Outputs digital signal (`True` or `False`) on a `pin`.
```

**Review by nkrkv — CHANGES_REQUESTED** [review]
Hope, these are the last changes

**Review by nkrkv — APPROVED** [review]
LGTM

**Pull request merged**